apps/remote_control/lib/lexical/remote_control/build.ex

```diff
@@ -35,6 +36,22 @@ defmodule Lexical.RemoteControl.Build do
     :ok
   end
 
+  def maybe_schedule_compile(triggered_file_uri, message) do
+    if Rename.in_progress?() do
+      Rename.update_progress(triggered_file_uri, message)
+    else
+      GenServer.cast(__MODULE__, {:compile, false})
+    end
+  end
+
+  def maybe_compile_document(%Project{} = project, %Document{} = document, message) do
+    if Rename.in_progress?() do
+      Rename.update_progress(document.uri, message)
+    else
+      compile_document(project, document)
+    end
+  end
```

Comment:
This sort of thing makes me a little nervous. There will be other commands/actions/etc. in the future that will take a long time and should potentially block compilation and adding additional checks here will get gross pretty fast.

So is the problem basically that, after issuing all these renames, we're going to get a bunch of messages from the client about all the files we just changed? And we don't want to actually compile until the client is done sending us all of those? I'd really like to consider whether there's a better, more general way to handle this sort of thing. For instance, does a similar thing happen when you checkout a different branch? Lexical gets a bunch of notifications about changed files, starts compiling mid-way through, and sometimes stuff ends up in a sort of borked state until you visit those files again?

This makes me think that compilation, perhaps, should be cancelable/restartable, and whenever a change comes in, it cancels any in-progress compilation and restarts it. So we might get a bunch of cancels/restarts, but now we'd have a system that is more resilient to rapid-fire changes to many files.

Reply:
> It's not just builds though, it's also formatting, code actions and the like

Yep, all the more reason to have some more general way of indicating that things are in a "partial" state. We definitely don't want every one of these places to have to know about every potentially long-running action.